### PR TITLE
Enable the coordinators to retrieve metadata from the catalog server [Experimental Feature]

### PR DESCRIPTION
**Pull request by erick576**
### Enable the coordinators to retrieve metadata from the catalog server
- Allowed a node to be discovered as a CatalogServer
- Added the catalog server module with its dependencies (nothing but the necessary ones and CatalogServer) and then installed them on the server module whenever the node is identified as a catalog server
- Allowed properties to be set for the catalog server

### Refrence
 - (https://github.com/prestodb/presto/pull/15479, https://github.com/prestodb/presto/pull/15673)

### Test plan
- You can try running **TestDiscoveryNodeManager.java** to see if the catalog server nodes are discoverable
- To test that the coordinators are retrieving metadata from the catalog servers you can try running a local query runner such as **TpchQueryRunner.java**. Just make sure to set the [coordinator count](https://github.com/prestodb/presto/blob/master/presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java#L86) to 1. Also change the node selection strategy on **TpchSplit.java** to have [NO_PREFRENCE](https://github.com/prestodb/presto/blob/master/presto-tpch/src/main/java/com/facebook/presto/tpch/TpchSplit.java#L89).
- Now you should be able to run and perform some queries. You can check the logs or debug to see that the catalog server is being called to retrieve the metadata. Ex : **select name from tpch.sf1.nation;**

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by zacw7** [comment]
The PR looks pretty solid. Thanks!

Let's break it into at least two PRs. (feel free to break them even smaller if you have some ideas)
PR1: 
1. Make ConnectorTransactionHandle mutable in TableHandle
2. Add RemoteMetadataManager implementation

PR2: 
1. Allow node to be discovered as a CatalogServer
2. Add CatalogServer and install module on server
3. Allow CatalogServer properties to be set

Also let's try not include irrelevant code change - like reformat the code to make it fit into 180 char per line limit. We can start another PR for that.

**Comment by erick576** [comment]
> > Sure will break the PR into two and regarding the line breakups should I undo them or just not put any more thats not a line I changed from now on
> 
> Let's undo them in our CatalogServer related PR. If you want we can start another PR for reformatting.

Has been split into 2 PRs. Undid all of the line breakdowns for lines that weren't apart of my changes. If I made a change to a line and the line became too long the change stands

**Comment by erick576** [comment]
Keep in mind, checks are failing because I had to split the PR into 2 and the config changes here need to inject the RemoteMetadataManager dependency which is now in the other PR. Please review #17907 first

**Comment by erick576** [comment]
> > Keep in mind, checks are failing because I had to split the PR into 2 and the config changes here need to inject the RemoteMetadataManager dependency which is now in the other PR #17907
> 
> It seems like `depends on` another PR is not working. How about we rebase this PR onto the other one to fix the failing checks. Once that PR is merged then rebase this PR onto master.

Sure, its rebased now

File: presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java [hunk 1]
```diff
@@ -523,6 +562,11 @@ private static boolean isResourceManager(ServiceDescriptor service)
...
     }
 
+    private static boolean isCatalogServer(ServiceDescriptor service)
+    {
+        return Boolean.parseBoolean(service.getProperties().get("catalog_server"));
+    }
+
     /**
      * The predicate filters out the services to allow selecting relevant nodes
      * for discovery and sending heart beat.
```
Comment (zacw7) [inline presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:572]: Let's add a comment for Catalog Server too.

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 2]
```diff
@@ -613,10 +613,15 @@ public ListeningExecutorService createResourceManagerExecutor(ResourceManagerCon
         // presto announcement
         checkArgument(!(serverConfig.isResourceManager() && serverConfig.isCoordinator()),
                 "Server cannot be configured as both resource manager and coordinator");
+        checkArgument(!(serverConfig.isResourceManager() && serverConfig.isCatalogServer()),
+                "Server cannot be configured as both resource manager and catalog server");
```
Comment (zacw7) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:634]: Do we have a strong reason for this?

File: presto-main/src/main/java/com/facebook/presto/catalogserver/RandomCatalogServerAddressSelector.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Function;
+
+import static com.facebook.drift.client.address.SimpleAddressSelector.SimpleAddress;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class RandomCatalogServerAddressSelector
```
Comment (zacw7) [inline presto-main/src/main/java/com/facebook/presto/catalogserver/RandomCatalogServerAddressSelector.java:33]: This class is almost identical to `RandomResourceManagerAddressSelector`. Is it possible to extract the common logic and share it with `ResourceManager` instead of copy-paste?

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java [hunk 4]
```diff
@@ -362,6 +377,17 @@ else if (resourceManager) {
             eventListenerManager = ((TestingEventListenerManager) injector.getInstance(EventListenerManager.class));
             clusterStateProvider = injector.getInstance(ResourceManagerClusterStateProvider.class);
         }
+        else if (catalogServer) {
+            dispatchManager = null;
+            queryManager = injector.getInstance(QueryManager.class);
```
Comment (zacw7) [inline presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java:382]: Why is `QueryManager` needed?

File: presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+        // presto coordinator announcement
+        discoveryBinder(binder).bindHttpAnnouncement("presto-catalog-server");
+
+        // failure detector
+        binder.install(new FailureDetectorModule());
+        jaxrsBinder(binder).bind(NodeResource.class);
+        jaxrsBinder(binder).bind(WorkerResource.class);
+        httpClientBinder(binder).bindHttpClient("workerInfo", ForWorkerInfo.class);
+
```
Comment (zacw7) [inline presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java:64]: Catalog Server shouldn't need to care about workers right?

File: presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        // query manager
+        binder.bind(QueryManager.class).to(NoOpQueryManager.class).in(Scopes.SINGLETON);
+
+        // resource group manager
+        binder.bind(ResourceGroupManager.class).to(NoOpResourceGroupManager.class);
```
Comment (zacw7) [inline presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java:69]: Why is this needed?

File: presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java [hunk 7]
```diff
@@ -45,32 +45,47 @@ public static DistributedQueryRunner createQueryRunner(Map<String, String> extra
         DistributedQueryRunner queryRunner = TpchQueryRunnerBuilder.builder()
                 .setExtraProperties(extraProperties)
                 .setResourceManagerEnabled(true)
+                .setCatalogServerEnabled(true)
```
Comment (zacw7) [inline presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java:48]: Let's leave this method as it is. Maybe create another method with a parameter `boolean catalogServerEnabled` to determine if it should be enabled or now.

**Review by zacw7 — CHANGES_REQUESTED** [review]
Overall looks good. Left a few comments.
The last commit `Allow Catalog properties to be set` can be merged into its previous one, so we can avoid that //TODO.

File: presto-main/src/main/java/com/facebook/presto/catalogserver/RandomCatalogServerAddressSelector.java (hunk 3 above)
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/catalogserver/RandomCatalogServerAddressSelector.java:33]: Is there any way to be able to have the resource manager and catalog server to use a different instance of this class? That would make having to use this method not needed

File: presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java (hunk 6 above)
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java:69]: Server requires that an resourceGroupManager is assigned upon startup or it complains. Thus it is assigned to the NoOp version which is basically the same thing as not having it at all

File: presto-main/src/main/java/com/facebook/presto/server/ServerConfig.java [hunk 8]
```diff
@@ -56,6 +58,30 @@ public ServerConfig setResourceManagerEnabled(boolean resourceManagerEnabled)
         return this;
     }
 
+    public boolean isCatalogServer()
+    {
+        return catalogServer;
+    }
+
+    @Config("catalog-server")
```
[the remark this answers is not in the document]
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/server/ServerConfig.java:66]: Added

File: presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java (hunk 5 above)
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java:64]: Thought they may of been needed, will remove

File: presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:572]: comment added

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java (hunk 4 above)
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java:382]: I did that since we assigned its query manager in the CatalogServerModule. Since its NoOp anyways I can probably put it to null and will have no effect

**Comment by erick576** [comment]
> Overall looks good. Left a few comments. The last commit `Allow Catalog properties to be set` can be merged into its previous one, so we can avoid that //TODO.

Squashed commit

**Comment by erick576** [comment]
> > Let's break it into at least two PRs. (feel free to break them even smaller if you have some ideas)
> 
> Is this PR still up for review? If we have broken up this, can we close this?

Yes, FYI this PR is already broken down. I've just stacked the 2 commits from #17907 to allow this to be tested

**Review by kewang1024 — COMMENTED** [review]
I don't think it's a great design to nest catalog server inside presto server, similar to presto-resource-group-manager, it should have its own repo instead of in presto-main
Thoughts? @zacw, @jainxrohit

**Comment by zacw7** [comment]
> I don't think it's a great design to nest catalog server inside presto server, similar to presto-resource-group-manager, it should have its own repo instead of in presto-main Thoughts? @zacw, @jainxrohit

I think for now it's ok. The functionality the current CatalogServer has is very minimal and not worth creating a separate module for it. Also, the code of [ResourceManager](https://github.com/prestodb/presto/tree/master/presto-main/src/main/java/com/facebook/presto/resourcemanager) (for multi-coordinator) are in `presto-main` as well. We would prefer keep the structure consistent.

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 9]
```diff
@@ -379,6 +385,11 @@ else if (serverConfig.isCoordinator()) {
                     }
                     return new ExceptionClassification(Optional.of(true), NORMAL);
                 });
+        binder.bind(RandomCatalogServerAddressSelector.class).in(Scopes.SINGLETON);
```
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:388]: Add a line break before and after this section

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 10]
```diff
@@ -539,7 +550,13 @@ public ListeningExecutorService createResourceManagerExecutor(ResourceManagerCon
         configBinder(binder).bindConfig(StaticFunctionNamespaceStoreConfig.class);
         binder.bind(FunctionAndTypeManager.class).in(Scopes.SINGLETON);
         binder.bind(MetadataManager.class).in(Scopes.SINGLETON);
-        binder.bind(Metadata.class).to(MetadataManager.class).in(Scopes.SINGLETON);
```
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:542]: Add a line break

File: presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+public class CatalogServerModule
+        extends AbstractConfigurationAwareModule
+{
+    @Override
+    protected void setup(Binder binder)
+    {
+        // discovery server
+        install(installModuleIf(EmbeddedDiscoveryConfig.class, EmbeddedDiscoveryConfig::isEnabled, new EmbeddedDiscoveryModule()));
+
+        // presto coordinator announcement
```
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java:54]: %s/coordinator/catalogServer

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 12]
```diff
@@ -539,7 +550,13 @@ public ListeningExecutorService createResourceManagerExecutor(ResourceManagerCon
...
+        if (serverConfig.isCatalogServerEnabled() && serverConfig.isCoordinator()) {
+            binder.bind(RemoteMetadataManager.class).in(Scopes.SINGLETON);
```
Comment (jainxrohit) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:554]: This should never be True, right? A server can not be both coordinator and catalogServer?

**Comment by jainxrohit** [comment]
> 

Lets discuss it. I am more leaning towards not adding too many things in the presto-main. Its becoming too fat.

**Review by zacw7 — COMMENTED** [review]
Please address Rohit's comments. Otherwise this PR looks pretty solid to me. Great job!

**Review by zacw7 — APPROVED** [review]

File: presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java [hunk 13]
```diff
@@ -51,23 +51,53 @@ public static DistributedQueryRunner createQueryRunner(Map<String, String> extra
         return queryRunner;
     }
 
-    public static DistributedQueryRunner createQueryRunner(Map<String, String> resourceManagerProperties, Map<String, String> coordinatorProperties, Map<String, String> extraProperties, int coordinatorCount)
+    public static DistributedQueryRunner createQueryRunner(Map<String, String> extraProperties, int coordinatorCount, boolean catalogServerEnabled)
```
Comment (jainxrohit) [inline presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java:54]: Are we using this method in any place?

**Review by jainxrohit — COMMENTED** [review]
Overall, I feel we should be able to run catalog server along with resource manager, so it should not be restricted it from our configurations. Right now our implementation is inspired from resource manager completely. We should discuss catalog server tradeoff separetely from the resource manager. 
Its okay to ship this PR from experiments perspective, but we should discuss in parallel to see what kind of choices we should make for catalog server.

**Comment by jainxrohit** [comment]
> perspective



> Overall, I feel we should be able to run catalog server along with resource manager, so it should not be restricted it from our configurations. Right now our implementation is inspired from resource manager completely. We should discuss catalog server tradeoff separetely from the resource manager. Its okay to ship this PR from experiments perspective, but we should discuss in parallel to see what kind of choices we should make for catalog server.

Adding @swapsmagic  for his opinion on this.

**Comment by zacw7** [comment]
> Overall, I feel we should be able to run catalog server along with resource manager, so it should not be restricted it from our configurations. Right now our implementation is inspired from resource manager completely. We should discuss catalog server tradeoff separetely from the resource manager. Its okay to ship this PR from experiments perspective, but we should discuss in parallel to see what kind of choices we should make for catalog server.

Thanks Rohit, actually we have considered that option.

Back when I had the discussions on the initial design with @yuanzhanhku and @NikhilCollooru, we actually made it clear that patching the functionality of `CatalogServer` on the `ResourceManager` was never the intention.

I also had an 1:1 with @swapsmagic when I was [prototyping](https://github.com/prestodb/presto/pull/17808) `CatalogServer` leveraging the `ResourceManager` setup, I believe our agreement was keeping them separate as well.

Back to your concern:
> we should be able to run catalog server along with resource manager, so it should not be restricted it from our configurations.

I agree with you that we should be able to run `CatalogServer` along with `ResourceManager` on the same physical machine and our current respects that philosophy, while by keeping them logically separate enable us to scale them separately. I personally do not think we should bundle them together.

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java (hunk 10 above)
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:542]: done

File: presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java (hunk 11 above)
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/server/CatalogServerModule.java]: Not sure I understand what you mean, do you want the comment to be clearer as to what we are announcing? Like replacing coordinator with "%s/coordinator/catalogServer"?

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java (hunk 12 above)
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:557]: No, isCatalogServerEnabled does not mean that the server is a catalogServer, that is isCatalogServer. What this check does is first see if the catalogServer is enabled at all meaning we are planning on actually using the catalog server and then if so the coordinator will make use of the remoteMetadataManager which is needed to call the catalog server instead of the standard one

File: presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java (hunk 13 above)
Comment (erick576) [reply presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java:54]: I originally had the catalogServer set on the createQueryRunner methods being used, but as @zacw7 recommended, I decided to remove them and leave it as is and introduce a method that would let the user, tester decide if they want to use this to enable the catalog server or not. Which is why this method is left for them to make use of or not

**Comment by jainxrohit** [comment]
> > Overall, I feel we should be able to run catalog server along with resource manager, so it should not be restricted it from our configurations. Right now our implementation is inspired from resource manager completely. We should discuss catalog server tradeoff separetely from the resource manager. Its okay to ship this PR from experiments perspective, but we should discuss in parallel to see what kind of choices we should make for catalog server.
> 
> Thanks Rohit, actually we have considered that option.
> 
> Back when I had the discussions on the initial design with @yuanzhanhku and @NikhilCollooru, we actually made it clear that patching the functionality of `CatalogServer` onto the `ResourceManager` was never the intention.
> 
> I also had an 1:1 with @swapsmagic when I was [prototyping](https://github.com/prestodb/presto/pull/17808) `CatalogServer` leveraging the `ResourceManager` setup, I believe our agreement was keeping them separate as well.
> 
> Back to your concern:
> 
> > we should be able to run catalog server along with resource manager, so it should not be restricted it from our configurations.
> 
> I agree with you that we should be able to run `CatalogServer` along with `ResourceManager` on the same physical node and our current design respects that philosophy, while by keeping them logically separate enable us to scale them separately. I personally do not think we should bundle them together. For example, there is an ongoing project to integrate Ratis to enable coordinator to select a master `ResourceManager`, which is not needed for `CatalogServer`; and scaling `CatalogServer` requires affinity which doesn't add much value for `ResourceManager` neither.

Yes, I only mean to have the flexibility to run both servers on the same physical machines. Some use-cases may require

File: presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java (hunk 13 above)
Comment (erick576) [reply presto-tests/src/test/java/com/facebook/presto/tests/tpch/TpchQueryRunner.java:54]: Sure, will remove for now then

File: presto-main/src/main/java/com/facebook/presto/catalogserver/RandomCatalogServerAddressSelector.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/catalogserver/RandomCatalogServerAddressSelector.java:33]: Not necessarily since the part that differers is here

>         List<HostAndPort> catalogServers = internalNodeManager.getCatalogServers().stream()

so there is no differing type between the two classes. Since they are both of type Set<InternalNode>. Their boolean "isResourceManager", etc in the InternalNode class will be what differs

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 14]
```diff
@@ -539,7 +552,14 @@ public ListeningExecutorService createResourceManagerExecutor(ResourceManagerCon
...
         binder.bind(MetadataManager.class).in(Scopes.SINGLETON);
-        binder.bind(Metadata.class).to(MetadataManager.class).in(Scopes.SINGLETON);
+
+        if (serverConfig.isCatalogServerEnabled() && serverConfig.isCoordinator()) {
+            binder.bind(RemoteMetadataManager.class).in(Scopes.SINGLETON);
+            binder.bind(Metadata.class).to(RemoteMetadataManager.class).in(Scopes.SINGLETON);
+        }
+        else {
+            binder.bind(Metadata.class).to(MetadataManager.class).in(Scopes.SINGLETON);
+        }
```
Comment (kewang1024) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:562]: Can we put it in next commit or PR? This is a change to coordinator instead of catalog however this commit is "install catalog server"

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 15]
```diff
@@ -613,8 +633,6 @@ public ListeningExecutorService createResourceManagerExecutor(ResourceManagerCon
         // presto announcement
         checkArgument(!(serverConfig.isResourceManager() && serverConfig.isCoordinator()),
                 "Server cannot be configured as both resource manager and coordinator");
-        checkArgument(!(serverConfig.isResourceManager() && serverConfig.isCatalogServer()),
-                "Server cannot be configured as both resource manager and catalog server");
```
Comment (kewang1024) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java]: Why are we removing this check?
If we don't need it, we shouldn't add it in the first commit

**Review by jainxrohit — APPROVED** [review]

**Review by kewang1024 — CHANGES_REQUESTED** [review]
1. Can you fix the PR title and description?
2. Looks like we're taking resource group feature as an example (https://github.com/prestodb/presto/pull/15479, https://github.com/prestodb/presto/pull/15673) then similarly can we also mark this feature as experimental and also leave a todo to move this component to its own dedicated directory after it's stabilized

File: presto-main/src/test/java/com/facebook/presto/metadata/TestDiscoveryNodeManager.java [hunk 16]
```diff
@@ -242,6 +273,18 @@ public void testGetResourceManagers()
         }
     }
 
+    @Test
+    public void testGetCatalogServers()
+    {
+        DiscoveryNodeManager manager = new DiscoveryNodeManager(selector, workerNodeInfo, new NoOpFailureDetector(), Optional.of(host -> false), expectedVersion, testHttpClient, new TestingDriftClient<>(), internalCommunicationConfig);
```
[the remark this answers is not in the document]
Comment (erick576) [reply presto-main/src/test/java/com/facebook/presto/metadata/TestDiscoveryNodeManager.java:279]: Is that possible? DiscoveryNodeManager is not of AutoCloseable type so putting it in the try parameters throws an error as it cannot be closed.

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java (hunk 15 above)
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java]: Commits will be squashed so this should be gone

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java (hunk 14 above)
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:562]: Ive squashed the commit so the message might apply to it now, if it still dosent make sense to have it on the main commit I can make another one for it

**Comment by erick576** [comment]
> Lets squash the commits, fix the PR description as well.

Squashed

File: presto-main/src/main/java/com/facebook/presto/metadata/InternalNode.java [hunk 17]
```diff
@@ -154,12 +164,19 @@ public boolean isResourceManager()
...
     public NodeVersion getNodeVersion()
     {
         return nodeVersion;
     }
 
-    @ThriftField(7)
+    @ThriftField(8)
     public NodeStatus getNodeStatus()
     {
         return nodeStatus;
```
[the remark this answers is not in the document]
Comment (erick576) [reply presto-main/src/main/java/com/facebook/presto/metadata/InternalNode.java:182]: The reasoning behind it was due to me adding a new thrift id between the last one which was why the ordering was pushed. If there is an effect I can append the new thrift id without adjusting existing ones

**Comment by zacw7** [comment]
Let's separate the commit message subject from body with a blank line following the [Chris Beams commit guidelines](https://chris.beams.io/posts/git-commit/).

**Comment by erick576** [comment]
> Let's separate the commit message subject from body with a blank line following the [Chris Beams commit guidelines](https://chris.beams.io/posts/git-commit/).

Strange I have line breaks between them on interactive mode and yet it isnt being shown on the commit message on github

**Comment by zacw7** [comment]
> > Let's separate the commit message subject from body with a blank line following the [Chris Beams commit guidelines](https://chris.beams.io/posts/git-commit/).
> 
> Strange I have line breaks between them on interactive mode and yet it isnt being shown on the commit message on github

Oh maybe it's just a UI issue.

**Comment by erick576** [comment]
> > > Let's separate the commit message subject from body with a blank line following the [Chris Beams commit guidelines](https://chris.beams.io/posts/git-commit/).
> > 
> > 
> > Strange I have line breaks between them on interactive mode and yet it isnt being shown on the commit message on github
> 
> Oh maybe it's just a UI issue.

I've been trying to add extra line breaks but the commit message just automatically truncates the new empty line away

File: presto-main/src/main/java/com/facebook/presto/metadata/InternalNode.java [hunk 18]
```diff
@@ -154,6 +164,13 @@ public boolean isResourceManager()
         return resourceManager;
     }
 
+    @ThriftField(8)
+    @Override
+    public boolean isCatalogServer()
+    {
+        return catalogServer;
+    }
+
```
Comment (kewang1024) [inline presto-main/src/main/java/com/facebook/presto/metadata/InternalNode.java:173]: NIT: can you move it after @ThriftField(7)

**Review by kewang1024 — APPROVED** [review]
1. Let's fix the PR description (remove the 3 commits since it's no longer true)
2. Although I prefer the old format (3 commits instead of squash as one), but since we have already made the change, I don't have strong opinion of changing it back

File: presto-main/src/main/java/com/facebook/presto/metadata/AllNodes.java [hunk 19]
```diff
@@ -27,14 +27,22 @@
     private final Set<InternalNode> shuttingDownNodes;
     private final Set<InternalNode> activeCoordinators;
     private final Set<InternalNode> activeResourceManagers;
+    private final Set<InternalNode> activeCatalogServers;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/metadata/AllNodes.java:30]: Are we sure we wanna make catalog server a separated tier from the beginning. Is it possible to make the module part of the resource manager as a first step? Having too many tiers will make our topology quite bit complicated.

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 20]
```diff
@@ -269,6 +272,9 @@ protected void setup(Binder binder)
         if (serverConfig.isResourceManager()) {
             install(new ResourceManagerModule());
         }
+        else if (serverConfig.isCatalogServer()) {
+            install(new CatalogServerModule());
+        }
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:277]: same comment as above. What if we wanna colocate resource manager and catalog server in the same node?

**Review by highker — APPROVED** [review]
